### ISSUE TITLE
Non-mapped mixin classes break ORM mapping

### ISSUE THREAD

**Issue by tlandschoff-scale** [sqlalchemy/sqlalchemy#4699]
Hi again,

it seems like I actually made things worse for us when reporting #4321. Our code actually mixes behaviour into our classes between the base class and the final concrete class. Nothing is mapped in those classes (truth to be told: one mixin class add relationships but from the ORM viewpoint we are still talking about single inheritance).

Here is a trivial example:

```
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import *


Base = declarative_base()


class User(Base):
    __tablename__ = 'user'

    id = Column(Integer, primary_key=True)
    name = Column(String)
    surname = Column(String)


class SpecialUser(User):
    # Marker Interface only
    pass



class ConvenienceStuff(User):
    def fullname(self):
        return self.name + " " + self.surname


class Manager(SpecialUser, ConvenienceStuff, User):
    __tablename__ = 'manager'

    id = Column(Integer, ForeignKey('user.id'), primary_key=True)
    title = Column(String)
```

Running this with SQLAlchemy 1.3.4 I get this exception:

```
sqlalchemy.exc.InvalidRequestError: Class <class '__main__.Manager'> has multiple mapped bases: [<class '__main__.SpecialUser'>, <class '__main__.ConvenienceStuff'>, <class '__main__.User'>]
```

This seems incorrect to me given that there are no additions mapping-wise in `ConvenienceStuff` and `SpecialUser`.

Originally we didn't have the `User` base class BTW, but this is incorrent both because any IDE will claim that `fullname` does not have access to attributes `name` and `surname`. Also, the method resolution order can be messed up with this.

I tried adding `__abstract__ = True` to those mix in classes but this gives a rather weird exception that's not better in any way:

```
Traceback (most recent call last):
  File "decl_example.py", line 31, in <module>
    class Manager(SpecialUser, ConvenienceStuff, User):
  File "/opt/scalesdk/python27/lib/python2.7/site-packages/sqlalchemy/ext/declarative/api.py", line 75, in __init__
    _as_declarative(cls, classname, cls.__dict__)
  File "/opt/scalesdk/python27/lib/python2.7/site-packages/sqlalchemy/ext/declarative/base.py", line 131, in _as_declarative
    _MapperConfig.setup_mapping(cls, classname, dict_)
  File "/opt/scalesdk/python27/lib/python2.7/site-packages/sqlalchemy/ext/declarative/base.py", line 160, in setup_mapping
    cfg_cls(cls_, classname, dict_)
  File "/opt/scalesdk/python27/lib/python2.7/site-packages/sqlalchemy/ext/declarative/base.py", line 192, in __init__
    self._setup_inheritance()
  File "/opt/scalesdk/python27/lib/python2.7/site-packages/sqlalchemy/ext/declarative/base.py", line 574, in _setup_inheritance
    "Class %s has multiple mapped bases: %r" % (cls, inherits)
sqlalchemy.exc.InvalidRequestError: Class <class '__main__.Manager'> has multiple mapped bases: [<class '__main__.User'>, <class '__main__.User'>, <class '__main__.User'>]
```

Switching back to SQLAlchemy release 1.1.18 this works fine again (as the changes for #4321 were not merged into 1.1).

Greetings, Torsten

**Comment by zzzeek** [comment]
my initial impression is the mixins should not be extending from the declarative hierarchy.  it's true you need the `__abtract__` flag if you want to have an intermediary class that isn't mapped, but this still does not support multiple inheritance right now.   any class that isn't mapped is better as a mixin that does not inherit from the mapped hierarchy at all, e.g. both SpecialUser and ConvenienceStuff.   if there is functionality that they both rely on on their superclass, then you make a base class like UserStuff that is also not descending from any mapped base.   But you've noted SpecialUser as a "marker interface", so that definitely has no concrete superclass from an OO perspective at least.

e.g.:
```

from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import *


Base = declarative_base()


class User(Base):
    __tablename__ = 'user'

    id = Column(Integer, primary_key=True)
    name = Column(String)
    surname = Column(String)


class SpecialUser(object):
    # Marker Interface only
    pass



class ConvenienceStuff(object):
    def fullname(self):
        return self.name + " " + self.surname


class Manager(SpecialUser, ConvenienceStuff, User):
    __tablename__ = 'manager'

    id = Column(Integer, ForeignKey('user.id'), primary_key=True)
    title = Column(String)

```

now can we add some more dragons or maybe a quick section on what `__abstract__` means to https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/mixins.html#declarative-mixins ?  absolutely.

**Comment by zzzeek** [comment]
the dragon would be, "mixins should not descend from a mapped class".   all of the examples at https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/mixins.html#declarative-mixins show this but i dont see that it's explicitly stated so perhaps that would be helpful here.

**Comment by tlandschoff-scale** [comment]
I am not sure how much work it is to redesign our class hierarchy for this. I'd like to illustrate with a (unfortunately) more complex example before you decide how to go about this.

But I am not sure if this is the right place for discussions, should I write to the mailing list instead or keep filling this ticket here?

Sorry for creating this scenario, but we were using classical mappings we did not restrict our usage of Python inheritance.

**Comment by zzzeek** [comment]
we can continue here, it's fine, github makes this way better than it used to be.

step 1 for moving to declarative would be reworking the inheritance patterns so that all mixins are not descending from concrete classes.

It's not as though declarative couldn't somehow be changed to accommodate for the pattern you're doing but it might make it harder for it to detect other kinds of mis-configurations and this would also require upstream development time.   The main one is that SQLAlchemy ORM mappings don't do multiple inheritance and declarative is trying to detect that here.

**Comment by tlandschoff-scale** [comment]
tl;dr: I think I whipped something up that I can apply to our application code base. I you wouldn't mind I'd be glad if you looked at it if this is a viable approach, @zzzeek. Feel free to close this ticket if you lack the time...

----
I spent the bulk of my last two work days to derive a toy example from our schema and move that from classical mapping to the modern mapping using the declarative extension.

I think I found a workable solution with only a small restriction left. Perhaps you can provide a hint for that? Namely I would like to enforce a fixed order for columns provided by a mixin class.

The code is too long for the issue tracker itself. I created a GitHub repository for the toy example.

* Original code (classical ORM mapping): https://github.com/tlandschoff-scale/sqlalchemy-gh4699/blob/original-classical-mapping/mixin_overrides.py
* Refined code (declarative ORM mapping): https://github.com/tlandschoff-scale/sqlalchemy-gh4699/blob/final-declarative-mapping/mixin_overrides.py

To make the code relatable I modelled a filesystem supporting files, directories and executables as special files. Directories and executables have support for "resource forks".

@zzzeek if it's not too much could you please take a look at the diff if the conversion makes sense to you? Perhaps you have an idea how to improve the things that seem awkward to me:

* Having both EntryInterface (abstract, non-mapped) and EntryCommon. I expect that our application will have a much bigger footprint for the interface class. :-(

* Marking each overriden setter for `_filesystem` as `@synonym_for`, this looks like something that can easily be forgotten. I am not really fond of properties anymore, but as I want to make querying available the same way as setting the field it really has to be a property. Maybe I could define the property only at the `EntryCommon` level and direct it to a method `_set_filesystem` to override the setter there.

I actually feel quite fond with the way the conversion went, the diff is here: https://github.com/tlandschoff-scale/sqlalchemy-gh4699/compare/original-classical-mapping...final-declarative-mapping

**Comment by zzzeek** [comment]
as you may have guessed I am holding off on reviewing this until I have a lot more time, as it seems you have solved your problem....as I probably mentioned earlier, I'm looking to change how mapper() works at a fundamental level so I should try to take a look at your recipes before then.

**Comment by tlandschoff-scale** [comment]
Hi Mike, you're right, by using the toy project as a guide we actually migrated away from using the classical mapper.

So take your time. :-)

Thanks, Torsten

**Comment by zzzeek** [comment]
in #5508, which is merging soon, declarative can now work without using a base at all.   maybe this helps:

https://gerrit.sqlalchemy.org/plugins/gitiles/sqlalchemy/sqlalchemy/+/refs/changes/02/2202/25/doc/build/orm/mapping_styles.rst#99

**Comment by zzzeek** [comment]
I was certain the new refactor I have for declarative would have fixed this, and it doesnt and it's a one liner.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Accommodate for same base class multiple times in inherits list** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2266
